### ISSUE TITLE
3rd time is a charm: support the new FreeDesktop standard dark/light theme preference (implemented by elementaryOS's Pantheon, GNOME 42, and soon KDE)

### ISSUE THREAD

**Issue by nekohayo** [lwindolf/liferea#1099]
_Finally_ there is a standardized way for Liferea to authoritatively know, without just trying to "guess" whether the current GTK theme is dark or not, what the user's intent is in terms of light/dark themes.

* Overview here: https://blogs.gnome.org/alexm/2021/10/04/dark-style-preference/
* Implementation tips here: https://gitlab.gnome.org/GNOME/Initiatives/-/wikis/Dark-Style-Preference

Knowing what the user actually _wishes_ for would let Liferea switch between the two in realtime with a clear standardized signal, and may also it confidently ship its own harmonized colors/stylesheet for both light/dark scenarios; for example, enforcing a white background on webpages when needed, and/or passing the information to webkit (like Epiphany and Firefox do), or being able to have two sets of colors for the "unread counters" in the list of feeds (ex: in light mode it could have a light gray background with black text, and in night mode it could have a dark gray background with a lighter/whiter shade of gray for counters' numbers' fonts).

You could let Liferea inherit the light vs dark status from the OS if the user wishes it to reflect that, or the user could decide that they want Liferea to remain light at all times for instance (maybe some folks will say that it's hard to read white-on-black, I dunno). In terms of UI design for handling those possibilities, yours is probably a rather simple one: "Follow system's preference", "Always light", or "Always dark", since I doubt you intend to offer multiple color schemes for each like IDEs/text editors do (but if you do want to go through that complicated route, https://github.com/settings/appearance is actually a pretty great example of such a preferences UI,  they seem to have solved the design problem of allowing both customized automatic and fully manual preferences).

Hopefully, with this cross-desktop standard, this time is the last time @lwindolf will have to mess with light vs dark themes, and it should be much easier and more reliable afterwards :) at least for elementary, GNOME [and KDE](https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/merge_requests/49). As for XFCE and others, I have no idea how far they are along for this.

------------

This might also help solve issues such as:

* #805
* #1031
* #1096

**Comment by nekohayo** [comment]
Oh and while we're at it, it might even allow you to confidently pass down the info to WebKitGTK (if it doesn't guess that automatically?) so that websites that check for it using CSS like `@media (prefers-color-scheme: dark) { blah blah }` can automagically appear with their correct tailored dark stylesheet in Liferea's built-in browser... You might be able to steal some code from the GNOME Web browser (Epiphany) as an example, because it already supports all of this.

For easier testing: some news websites (besides social networks and geeky websites like GitHub / Matrix / Discourse) that support it are 9to5linux.com and foreignaffairs.com, they show up with a dark style in Firefox's dark mode.

**Comment by lwindolf** [comment]
72e9567b adds prefer dark mode support. Which is needed for this,

**Comment by lwindolf** [comment]
Works quite well. With the exception of a small problem: right after dark mode switching the HTML pane is inverse colored. Once another item is clicked it is correct. Very strange...

**Comment by lwindolf** [comment]
I've tested automatic dark theme switching with foreignaffairs.com which works. 9to5linux.com does not seem to support anymore though (tested with Firefox).

**Comment by lwindolf** [comment]
I consider this now implemented. I will track the remaining issue of the HTML pane "lagging" one click behind with the colors with #1031

**Comment by nekohayo** [comment]
I didn't see Liferea 1.14.0 showing up darkly when launched from the flathub flatpak today, while my GNOME 43 desktop was in dark mode already. As there is no setting for this, I'm wondering if there might be a bug; is there something I need to be looking out for in particular?

**Comment by lwindolf** [comment]
@nekohayo I did not test in a flathub setup (just do not have the time). It works for me with GNOME without a portal.

For debugging you can run

      liferea --debug-conf | grep "FDO reports color-schema code"

If this debug message does not appear it means the portal does not support exposing the color-scheme key. If it does it should say value 1 for non-dark, 0 or 2 for dark.

**Comment by bbhtt** [comment]
Is this reading the `org.gnome.desktop.interface` gsettings key to get the preference? Flatpak doesn't allow accessing those https://gitlab.gnome.org/GNOME/glib/-/issues/2213 and inside the sandbox they are always set to the default values. 

So if  `$ gsettings get org.gnome.desktop.interface color-scheme` reports `'prefer-dark'` on host, inside the sandbox it'll always be set to `default`: 

```
$ flatpak run --command=sh net.sourceforge.liferea

gsettings get org.gnome.desktop.interface color-scheme
'default'
```

Changing this on host with GNOME settings>Appearance or other methods won't affect this value. With `--debug-conf`  I can see that Liferea is always reporting `default` even if I set dark in GNOME Settings>Appearance.

The correct way I think is to monitor `org.freedesktop.appearance` over dbus Examples. [1](https://github.com/kovidgoyal/kitty/commit/c2bc24de35873f143c00523d51fbc5fa2ad9779f), [2](https://hg.mozilla.org/mozilla-central/file/e5ed236608191461d74f0b5a291d57d74657ddab/widget/gtk/nsLookAndFeel.cpp#l1352)

**Comment by nekohayo** [comment]
This ticket might need to be reopened as per the bbhtt's comment above?

**Comment by lwindolf** [comment]
@bbhtt I do agree on the brokeness for Flatpak. I do not see a solution yet though. This is not maintainable for a small project not using libadwaita. This is another case of GNOME accidently pushing to get rid of GTK-only projects. Those guys building this type of features (Flatpak portals) in fancy high level abstractions have simply seem to have no clue on the impact on the non-GNOME ecosystem. How hard would it be for them to provide a simple wrapper library?

End of rant

**Comment by nekohayo** [comment]
Hmm, something seems amiss, still... I tested with the **non-flatpak** (RPM) version 1.14.4 from Fedora 38 (I guess 1.14.5 will trickle down soon enough, but I didn't see something related to this in the changelog anyway) and even there, the FreeDesktop dark color scheme doesn't seem to be fully applied in practice by Liferea, whether in live changes or on restart of the app. What it does is that it changes the colors of the unread items counters in the sidebar, but it doesn't monitor for those changes to be applied in realtime (only on app restart) and doesn't apply a dark variant of the GTK theme either.

`gsettings get org.gnome.desktop.interface color-scheme` prints `'prefer-dark'`, and `liferea --debug-conf | grep "color-schema"` prints:

```
CONF   : GNOME reports color-schema 'prefer-dark'
CONF   : GNOME reports color-schema 'prefer-dark'
CONF   : GNOME reports color-schema 'prefer-dark'
```

For what it's worth, opening the [GTK Inspector](https://blog.gtk.org/2017/04/05/the-gtk-inspector/) from Liferea, going into the Visuals (Appearance?) tab and turning on the "Dark variant" switch works, but I think (if I'm not mistaken) that it's GTK3-specific.

To summarize: yes, to a certain extent Liferea is _aware_ of the freedesktop dark mode on startup, just not in realtime and just not applying the gtk theme change...

I hope this info helps somehow.
